File: everything_parsing/13-subset-construction/java/Assert.java
// Everything Parsing — the whole test framework, on purpose.
// Canonical copy: prerequisites/assert/Assert.java
// Each module keeps its own copy in java/ so `javac java/*.java` just works.
//
// No JUnit, no Gradle. A test file is a class with a `public static void main`
// that calls these and finishes with Assert.summary(). Non-zero exit on failure
// so `run.md` commands and CI can tell pass from fail.

public final class Assert {
    private static int passed = 0;
    private static int failed = 0;

    private Assert() {}

    public static void that(boolean cond, String label) {
        if (cond) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAIL  " + label);
        }
    }

    public static void equals(Object actual, Object expected, String label) {
        boolean ok = (actual == null) ? expected == null : actual.equals(expected);
        if (ok) {
            passed++;
        } else {
            failed++;
            System.out.println("  FAIL  " + label);
            System.out.println("        expected: " + show(expected));
            System.out.println("        actual  : " + show(actual));
        }
    }

    private static String show(Object o) {
        if (o == null) return "null";
        String s = o.toString();
        return s.contains("\n") ? "\n----\n" + s + "\n----" : "\"" + s + "\"";
    }

    /** Print the tally and exit non-zero if anything failed. Call last. */
    public static void summary() {
        System.out.println();
        System.out.println(failed == 0
            ? "OK   " + passed + " passed"
            : "FAIL " + failed + " failed, " + passed + " passed");
        if (failed != 0) System.exit(1);
    }
}

File: everything_parsing/02-strings-languages-operations/js/assert.mjs
// Everything Parsing — the whole test framework, on purpose.
// Canonical copy: prerequisites/assert/assert.mjs
// Each module keeps its own copy in js/ so imports stay relative and simple.
//
// No test runner. A *.test.mjs file imports these, calls them, and ends with
// summary(). process.exit(1) on failure so run.md commands and CI can tell.

let passed = 0;
let failed = 0;

export function that(cond, label) {
  if (cond) { passed++; return; }
  failed++;
  console.log("  FAIL  " + label);
}

export function equals(actual, expected, label) {
  const a = norm(actual);
  const e = norm(expected);
  if (a === e) { passed++; return; }
  failed++;
  console.log("  FAIL  " + label);
  console.log("        expected: " + show(e));
  console.log("        actual  : " + show(a));
}

function norm(v) {
  if (typeof v === "string") return v;
  if (typeof v === "bigint") return v.toString() + "n";
  return JSON.stringify(v);
}

function show(s) {
  return String(s).includes("\n") ? "\n----\n" + s + "\n----" : '"' + s + '"';
}

export function summary() {
  console.log();
  console.log(failed === 0
    ? "OK   " + passed + " passed"
    : "FAIL " + failed + " failed, " + passed + " passed");
  if (failed !== 0) process.exit(1);
}
